### PR TITLE
[s] Fixes signaler exploit

### LINKED ISSUE #45855: you can mind transfer while signaler suiciding
https://cdn.discordapp.com/attachments/326826573951008769/609973634240610305/suicide_wiz_3.PNG

### PR DESCRIPTION
**Pull request by nemvar**
## About The Pull Request

This prevents wizards from being epic gamers. Fixes #45855

## Changelog
:cl:
fix: Signaler suicide is now bound to the mind, instead of the body.
/:cl:

### CODE REVIEW

**Comment by Tlaltecuhtli** [comment]
what if the wizard makes meth bombs with long prime timer and mindswaps after suicide? what about other not instant suicides like nuke disk

isnt it better to make suiciders just unable to mindswap after?

**Comment by KomradeSpectre** [comment]
Just want to note that this potentially is still an issue with any other delayed suicide, all I have is data from the round mentioned in that report.

**Comment by nemvar** [comment]
The problem with this is that its suicide. The suicide verb should not and should never give you gameplay advantage. It exists so you can remove yourself from the round in a fun way.

**Comment by KomradeSpectre** [comment]
Yeah, I'm completely aware of this and fully agree. See the screenshot listed in the issue report where a user (TheMythicGhost), me, rages about it despite not being a victim of it. Just thought I'd give the heads up just in case it's extended to other things as well, and those potentially would be at risk of recreating this situation.

**Comment by nemvar** [comment]
@Tlaltecuhtli No idiot. Suicide should always be on the mind. It's a shame that we even track suicide on the mob in the first place and it should be migrated over completely.

**Comment by ATH1909** [comment]
> isnt it better to make suiciders just unable to mindswap after?

They're already not supposed to be able to, but the signaler suicide is weird and doesn't actually set your "suiciding" variable to TRUE until the signaler is pulsed. I believe that this is because of https://github.com/tgstation/tgstation/pull/43036.

File: code/modules/assembly/signaler.dm [hunk 1]
```diff
@@ -13,22 +13,34 @@
...
-/obj/item/assembly/signaler/proc/manual_suicide(mob/living/carbon/user)
-	user.visible_message("<span class='suicide'>[user]'s [src] receives a signal, killing [user.p_them()] instantly!</span>")
+/obj/item/assembly/signaler/proc/manual_suicide(datum/mind/suicidee)
+	var/mob/living/user = suicidee.current
+	if(!istype(user))
+		return
+	if(suicide_mob == REF(user))
+		user.visible_message("<span class='suicide'>[user]'s [src] receives a signal, killing [user.p_them()] instantly!</span>")
+	else
+		user.visible_message("<span class='suicide'>[user]'s [src] receives a signal and dies like a gamer!</span>")
```
Comment (ATH1909) [inline code/modules/assembly/signaler.dm]: So this'll kill them in a special manner if they tried to do my trick themselves (after this PR is merged)? I approve.

**Comment by ATH1909** [comment]
> Yeah, I'm completely aware of this and fully agree. See the screenshot listed in the issue report where a user (TheMythicGhost), me, rages about it despite not being a victim of it. Just thought I'd give the heads up just in case it's extended to other things as well, and those potentially would be at risk of recreating this situation.

Wait, you're TheMythicGhost? So you not only repeatedly called for blood in OOC chat after the round despite not even dying to my gimmick, but you then went into my issue report (linked in the opening post of this PR) and tried to get me lynched (t)here too? What the hell, man?

File: code/modules/assembly/signaler.dm (hunk 1 above)
Comment (nemvar) [reply code/modules/assembly/signaler.dm]: Actually messed up the grammar here. Will fix later.

**Comment by ATH1909** [comment]
I just checked #43036 for other suicides that could be used with this trick, and I found only one: The coin suicide, although it'd be really, REALLY hard to pull this trick off with that suicide (you have only a ONE SECOND window in which to mind transfer out before you risk hoisting yourself by your own petard, and even if you pull the trick off perfectly, there's still a 50% chance that the coin suicide doesn't kill your old body/your victim).

**Comment by ATH1909** [comment]
This should with the timer suicide too (even though it wasn't modified by #14982), and with any suicide that has a delay between activating it and it returning something. Yikes.

EDIT: And even the saber suicide (and maybe the desword suicide too?)... I don't think that the code for mind transfer suicides does what it's intended to do, unless I'm misreading the code here.

This might be a bigger problem than we thought.

**Comment by ATH1909** [comment]
Wait, I'm just a dingus that doesn't know how to read code. The coin suicide is the only suicide other than the signaler one that this works with.

File: code/modules/assembly/signaler.dm [hunk 2]
```diff
@@ -13,22 +13,34 @@
 	var/frequency = FREQ_SIGNALER
 	var/delay = 0
 	var/datum/radio_frequency/radio_connection
-	var/suicider = null
+	///Holds the mind that commited suicide.
+	var/datum/mind/suicider = null
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/assembly/signaler.dm]:
```suggestion
	var/datum/mind/suicider
```

File: code/modules/assembly/signaler.dm [hunk 3]
```diff
@@ -13,22 +13,34 @@
...
+		return
+	if(suicide_mob == REF(user))
+		user.visible_message("<span class='suicide'>[user]'s [src] receives a signal, killing [user.p_them()] instantly!</span>")
+	else
+		user.visible_message("<span class='suicide'>[user]'s [src] receives a signal and [user.p_they()] die[user.p_s()] like a gamer!</span>")
 	user.adjustOxyLoss(200)//it sends an electrical pulse to their heart, killing them. or something.
 	user.death(0)
 	user.set_suicide(TRUE)
 	user.suicide_log()
+	playsound(user, 'sound/machines/triple_beep.ogg', ASSEMBLY_BEEP_VOLUME, TRUE)
```
Comment (ATH1909) [inline code/modules/assembly/signaler.dm]: wait, did the signaler not beep before when triggered as a part of a suicide attempt?

**Comment by ExcessiveUseOfCobblestone** [comment]
bro press commit merge and/or give maintainers access to your code smh.

File: code/modules/assembly/signaler.dm (hunk 3 above)
Comment (nemvar) [reply code/modules/assembly/signaler.dm]: they did, it's just that I nullspace the signaler.

**Comment by ExcessiveUseOfCobblestone** [comment]
living dangerously 8)

**Pull request merged**